File: neet_code_all/src/stack/validate_stack_sequences.rs
// 題意が理解できなかった
struct Solution;
impl Solution {
    pub fn validate_stack_sequences(pushed: Vec<i32>, popped: Vec<i32>) -> bool {
        true
    }
}

// Pythonの模範解答より
struct SolutionAnsPython;
impl SolutionAnsPython {
    // AC
    pub fn validate_stack_sequences(pushed: Vec<i32>, popped: Vec<i32>) -> bool {
        let mut idx = 0;
        let mut stack = vec![];
        for n in pushed {
            stack.push(n);

            while idx < popped.len() && stack.len() != 0 && popped[idx] == *stack.last().unwrap() {
                stack.pop();
                idx += 1;
            }
        }

        stack.len() == 0
    }
}

fn main() {
    let case_1 = (vec![1, 2, 3, 4, 5], vec![4, 5, 3, 2, 1]);
    // => true
    let case_2 = (vec![1, 2, 3, 4, 5], vec![4, 3, 5, 1, 2]);
    // => false

    println!(
        "case_1: {}",
        SolutionAnsPython::validate_stack_sequences(case_1.0.clone(), case_1.1.clone())
    );
    println!(
        "case_2: {}",
        SolutionAnsPython::validate_stack_sequences(case_2.0.clone(), case_2.1.clone())
    );
}

File: neet_code_all_cpp/stack/validate_stack_sequences.cpp
#include <iostream>
#include <vector>
#include <stack>

void show_stack(std::stack<int> stk) {
    while(!stk.empty()) {
        std::cout << stk.top() << "";
        stk.pop();
    }
    std::cout << std::endl;
}

// AC
class SolutionAnsPython {
    public:
        bool validateStackSequences(std::vector<int>& pushed, std::vector<int>& popped) {
            int idx = 0;
            std::stack<int> stack;
            for(auto& n: pushed){
                stack.push(n);
                while(idx < popped.size() && !stack.empty() && popped[idx] == stack.top()) {

                    stack.pop();
                    ++idx;
                }
            }
            //show_stack(stack);

            return stack.empty();
        }
};

int main(void) {
    using Case = std::pair<std::vector<int>, std::vector<int>>;
    Case case_1 = {{1, 2, 3, 4, 5}, {4, 5, 3, 2, 1}};
    // => true
    Case case_2 = {{1, 2, 3, 4, 5}, {4, 3, 5, 1, 2}};
    // => false

    SolutionAnsPython s_ans_py;

    std::cout << s_ans_py.validateStackSequences(case_1.first, case_1.second) << std::endl; 
    std::cout << s_ans_py.validateStackSequences(case_2.first, case_2.second) << std::endl; 
}
